### PR TITLE
Add a benchmark and additional tests.

### PR DESCRIPTION
**Pull request by veluca93**
Also add a script to help running sanitizer builds, and do some cleanup.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -72,4 +76,4 @@ jobs:
         with:
           path: ~/.cache/bazel
           key: bazel-${{ runner.os }}
-      - run: bazel build --cxxopt=-std=c++20 //...
+      - run: bazel build -c opt --cxxopt=-std=c++20 //...
```
Comment (jan-wassenberg) [inline .github/workflows/build.yml:79]: FYI we deliberately omitted opt to speed up builds. Internal CI has a ~7 min deadline and we are pretty close to exceeding that :/ Bazel is slower than CMake. Any thoughts on how else to speed up the build?

File: benchmark.cc [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+#include <vector>
+
+#include "hwy/base.h"
+#include "hwy/contrib/thread_pool/thread_pool.h"
+#include "hwy/highway.h"
+#include "hwy/per_target.h"
+#include "hwy/profiler.h"
+#include "hwy/timer.h"
+#include "nlohmann/json.hpp"
+#include "util/app.h"
```
Comment (jan-wassenberg) [inline benchmark.cc:16]: These two require the next_line comment.

File: benchmark.cc [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+#include "util/args.h"
+
+using json = nlohmann::json;
+
+class BenchmarkArgs : public gcpp::ArgsBase<BenchmarkArgs> {
+ public:
+  BenchmarkArgs(int argc, char* argv[]) { InitAndParse(argc, argv); }
+
+  std::string goldens_dir;
```
Comment (jan-wassenberg) [inline benchmark.cc:26]: Consider using Path for these?

File: benchmark.cc [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+void LogSpeedStats(const double time_start, size_t total_tokens) {
+  const double time_end = hwy::platform::Now();
+  const double time_elapsed = time_end - time_start;
+  const double tok_sec = total_tokens / time_elapsed;
+  std::cout << total_tokens << " tokens in " << time_elapsed << " seconds"
+            << " [" << tok_sec << " tokens / sec" << "]\n";
+}
+
+std::pair<std::string, int> query_model(
```
Comment (jan-wassenberg) [inline benchmark.cc:60]: PascalCase naming please :)

File: benchmark.cc [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    return true;
+  };
+  if (app.verbosity >= 2) {
+    std::cout << args.max_tokens << " " << args.max_generated_tokens << " "
+              << args.temperature;
+  }
+  GenerateGemma(model, args.max_tokens, args.max_generated_tokens,
+                args.temperature, prompt, /*abs_pos=*/0, kv_cache, pool,
+                inner_pool, stream_token, accept_token, gen, app.verbosity);
+  // const double time_end = hwy::platform::Now();
```
Comment (jan-wassenberg) [inline benchmark.cc:95]: Remove? This seems to be in LogSpeedStats.

File: benchmark.cc [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+      std::cout << "WRONG\n\n";
+    }
+    if (++i >= max_questions) break;
+  }
+  printf("Correct answers: %zu / %zu\n", correct_answers, i);
+  return EXIT_SUCCESS;
+}
+
+/* Run this in the same way as gemma, p.ex.:
+ ./benchmark --tokenizer tokenizer.spm --model 2b-it --compressed_weights \
```
Comment (jan-wassenberg) [inline benchmark.cc:253]: --weights is now preferred, right?

File: gemma.cc [hunk 7]
```diff
@@ -1090,5 +1235,16 @@ void CompressWeights(gcpp::Model model, const Path& weights,
   (model, weights, compressed_weights, pool);
 }
 
+float ComputeCrossEntropy(Gemma& gemma, size_t max_tokens,
+                          const std::vector<int>& prompt, KVCache& kv_cache,
+                          hwy::ThreadPool& pool, hwy::ThreadPool& inner_pool,
+                          int verbosity) {
+  pool.SetWaitMode(hwy::PoolWaitMode::kSpin);
```
Comment (jan-wassenberg) [inline gemma.cc:1242]: Is this time-critical? We can stick with the default/current wait mode, I think.

File: gemma_test.cc [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+// limitations under the License.
+
+// copybara:import_next_line:gemma_cpp
+#include "gemma.h"
+
+#include <thread>
+
+#include "gtest/gtest.h"
+#include "hwy/contrib/thread_pool/thread_pool.h"
+#include "ops.h"
```
Comment (jan-wassenberg) [inline gemma_test.cc:23]: Also requires next_line.

**Review by jan-wassenberg — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yml (hunk 1 above)
Comment (veluca93) [reply .github/workflows/build.yml:79]: No strong opinions :-)

Comment (jan-wassenberg) [reply .github/workflows/build.yml:79]: Let's remove opt then for now.

File: gemma.cc (hunk 7 above)
Comment (veluca93) [reply gemma.cc:1242]: It's part of a benchmark, which I assume why this is here :-)

**Review by jan-wassenberg — APPROVED** [review]

**Pull request merged**